src/mito2/src/worker/handle_catchup.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Handling catchup request.
+
+use std::sync::Arc;
+
+use snafu::ensure;
+use store_api::logstore::LogStore;
+use store_api::region_request::{AffectedRows, RegionCatchupRequest};
+use store_api::storage::RegionId;
+
+use crate::error::{self, Result};
+use crate::region::opener::{replay_memtable, RegionOpener};
+use crate::worker::RegionWorkerLoop;
+
+impl<S: LogStore> RegionWorkerLoop<S> {
+    pub(crate) async fn handle_catchup_request(
+        &mut self,
+        region_id: RegionId,
+        request: RegionCatchupRequest,
+    ) -> Result<AffectedRows> {
+        let Some(region) = self.regions.get_region(region_id) else {
+            return error::RegionNotFoundSnafu { region_id }.fail();
+        };
+
+        if region.is_writable() {
+            return Ok(0);
+        }
+        // Note: Currently, We protect the split brain by ensuring the mutable table is empty.
+        // It's expensive to execute catch-up requests without `set_writable=true` multiple times.
+        let is_mutable_empty = region.version().memtables.mutable.is_empty();
+
+        let region = if !is_mutable_empty || region.manifest_manager.check_update().await? {
+            let reopened_region = Arc::new(
+                RegionOpener::new(
+                    region_id,
+                    region.region_dir(),
+                    self.memtable_builder.clone(),
+                    self.object_store_manager.clone(),
+                    self.scheduler.clone(),
+                )
+                .cache(Some(self.cache_manager.clone()))
+                .options(region.version().options.clone())
+                .open(&self.config, &self.wal)
```

Comment:
I think we already replay after `open()`. We can replay the memtable only in the `else` branch.

Reply:
I changed it to open with `skip_wal_replay = true.` I prefer to reuse the below `memtable_replay`.